io-sim/src/Control/Monad/IOSimPOR/Internal.hs

```diff
@@ -874,34 +978,62 @@ reschedule simstate@SimState{ threads, timers, curTime = time, races } =
+        (wakeupSTM, wokeby) <- threadsUnblockedByWrites written
         mapM_ (\(SomeTVar tvar) -> unblockAllThreadsFromTVar tvar) written
 
-        -- TODO: the vector clock below cannot be right, can it?
-        let (unblocked,
-             simstate') = unblockThreads bottomVClock wakeup simstate
-            -- all open races will be completed and reported at this time
-            simstate''  = simstate'{ races = noRaces }
+        let wakeupThreadDelay = [ tid | TimerThreadDelay tid <- fired ]
+            wakeup            = wakeupThreadDelay ++ wakeupSTM
+            -- TODO: the vector clock below cannot be right, can it?
+            (_, !simstate')   = unblockThreads bottomVClock wakeup simstate
+
+            -- For each 'timeout' action where the timeout has fired, start a
+            -- new thread to execute throwTo to interrupt the action.
+            !timeoutExpired = [ (tid, tmid, isLockedRef)
+                              | TimerTimeout tid tmid isLockedRef <- fired ]
+
+        -- Get the isLockedRef values
+        !timeoutExpired' <- traverse (\(tid, tmid, isLockedRef) -> do
+                                        locked <- readSTRef isLockedRef
+                                        return (tid, tmid, isLockedRef, locked)
+                                    )
+                                    timeoutExpired
+
+        -- all open races will be completed and reported at this time
+        !simstate'' <- forkTimeoutInterruptThreads timeoutExpired'
+                                                   simstate' { races = noRaces }
         !trace <- reschedule simstate'' { curTime = time'
                                         , timers  = timers' }
         let traceEntries =
-                     [ (time', ThreadId [-1], (-1), Just "timer", EventTimerExpired tmid)
-                     | tmid <- tmids ]
-                  ++ [ (time', tid', (-1), tlbl', EventTxWakeup vids)
-                     | tid' <- unblocked
+                     [ ( time', ThreadId [-1], -1, Just "timer"
+                       , EventTimerExpired tmid)
+                     | (tmid, Timer _) <- zip tmids fired ]
+                  ++ [ ( time', ThreadId [-1], -1, Just "register delay timer"
+                       , EventRegisterDelayExpired tmid)
+                     | (tmid, TimerRegisterDelay _) <- zip tmids fired ]
+                  ++ [ (time', tid', -1, tlbl', EventTxWakeup vids)
+                     | tid' <- wakeupSTM
                      , let tlbl' = lookupThreadLabel tid' threads
                      , let Just vids = Set.toList <$> Map.lookup tid' wokeby ]
+                  ++ [ ( time', tid, -1, Just "thread delay timer"
+                       , EventThreadDelayExpired)
+                     | tid <- wakeupThreadDelay ]
+                  ++ [ ( time', tid, -1, Just "timeout timer"
+                       , EventTimeoutFired tmid)
+                     | (tid, tmid, _, _) <- timeoutExpired' ]
```

Comment:
As in `IOSim` case, we ought to record thread started with `forkTimeoutInterruptThreads`.

Reply:
What's strange in this block is that we log:
* `EventTimerExpired`
* `EventRegisterDelayExpired`
* `EventThreadDelayExipred`
* `EventTimeoutFired`

The last one does not fit to the other ones, so something is wrong.  Were are we tracing:
* `EventTimerFired`
* `EventRegisterDelayFired`
* `EventThreadDelayFired`

Reply:
Should be here but I named them differently.